cdap-common/src/main/java/co/cask/cdap/common/namespace/DefaultNamespacedLocationFactory.java

```diff
@@ -33,23 +33,25 @@
  */
 public class DefaultNamespacedLocationFactory implements NamespacedLocationFactory {
 
-  private final CConfiguration cConf;
   private final LocationFactory locationFactory;
+  private final String namespaceDir;
 
   @Inject
   public DefaultNamespacedLocationFactory(CConfiguration cConf, LocationFactory locationFactory) {
-    this.cConf = cConf;
+    this.namespaceDir = cConf.get(Constants.Namespace.NAMESPACES_DIR);
     this.locationFactory = locationFactory;
   }
 
   @Override
   public Map<Id.Namespace, Location> list() throws IOException {
-    String namespacesDir = cConf.get(Constants.Namespace.NAMESPACES_DIR);
-    Location namespacesLocation = locationFactory.create(namespacesDir);
+    Location namespacesLocation = getNamespacesBaseLocation();
     Map<Id.Namespace, Location> namespaceLocations = Maps.newHashMap();
     if (namespacesLocation.exists()) {
       for (Location namespaceLocation : namespacesLocation.list()) {
-        namespaceLocations.put(Id.Namespace.from(namespaceLocation.getName()), namespaceLocation);
+        // Any directories started with "." is special system file, which is not regular namespace directory
+        if (!namespaceLocation.getName().startsWith(".")) {
+          namespaceLocations.put(Id.Namespace.from(namespaceLocation.getName()), namespaceLocation);
```

Comment:
Actually I found this implementation interesting. Why the listing of namespace locations not come from NamespaceAdmin, which should be consulting the namespace store and constructing the Location? What if someone just create a directory (a valid namespace name) under there? Then it becomes a valid namespace as returned by this class, but actually missing from the namespace store?